how-it-works/1_protocol/01_peer-to-peer-p2p.subpage.md

```diff
@@ -5,10 +5,73 @@ shareImage: /img/how-it-works/peer-to-peer-p2p.600.jpg
+## Sparse Overlay Network Topology
 
+In subnets up to a certain size, it is practical that each node be connected to each other node of the subnet in terms of the peer relationship, i.e., each node having all the other nodes as its peers.
+Full connectivity between the nodes trivially solves the problem of nodes being *eclipsed*, i.e., cut off from the broadcast network, by only connecting to compromised nodes.
+In large subnets, however, this full connectivity is neither practical nor desirable as it results in too large a number of messages being communicated for a single broadcast of an artifact.
 
+In the future, we will address this challenge by implementing an optimized solution such as an *overlay network* that is more sparse than the fully connected subnet. Broadcast through gossip is implemented in this overlay network instead of the fully connected subnet, thereby leading to a greatly reduced number of messages.
+The overlay network is randomly created with properties that ensure, with high probability, that nodes cannot be *eclipsed* by only being connected to adversarial nodes.
+The overlay network structure can be recomputed periodically in order to prevent adaptive eclipse attacks where an adversary over time corrupts an increasing number of nodes.
+
+## Implementation Architecture
+
+The P2P layer's implementation architecture uses the concept of *artifact pools* for nodes to keep track of received and to-be-broadcast artifacts.
+An artifact pool has a *validated* and an *unvalidated* section, containing artifacts that have been validated by the upper protocol layer of the node or generated on the node, or containing artifacts that are still unvalidated, respectively.
+Different pools are used for different kinds of artifacts.
+Whenever the P2P layer receives an artifact from another node, it becomes part of an unvalidated pool.
+P2P then requests its validation from components in a higher protocol layer, and once validated, the artifact is moved to the validated pool for this kind of artifact.
+Artifacts generated by the protocol on the node are immediately placed into a validated pool, as being generated by the protocol locally makes them implicitly validated.
+
+The implementation uses *adverts queues* to track received adverts, a *requested queue* to track which artifacts are currently being downloaded from peers, and a *receive check* data structure that tracks which artifacts have been downloaded already.
+Those data structures help realize an efficient advert-based gossip protocol supporting artifact priorities.
+
+Each node maintains multiple parallel open connections to each of its peers to better utilize the available network bandwidth.
```

Comment:
This is not true in the *current* implementation. It will become true hopefully in a few weeks.